stanfordnlp/dspy
# fix(dependencies): move builds and dependencies to poetry

## 📝 Changes Description
This PR contains the following changes:

- migrated setuptools to poetry
- remove poetry main project stuffs and move them to the tool.dependencies
- export the poetry stuffs to requirements to if someone wanted to use it till June
- update PR structure message.

## 🛠 Type of Change

- [ ] This change will expire our current cache and need cache update


## 🔍 How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Updated one notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

## How to test?

- the way that you can test it is to setup an environment and follow the steps for setting up your poetry and use that.

- we should run notebooks and make sure they works.

- another part that should be tested is the build of the package and push it to pypi, which should be done with poetry build 

- To publish a Python package to PyPI using Poetry, you need to follow these steps:

1. Poetry will create a distribution package that you can upload to PyPI. You can build the package with the following command:

```bash
poetry build
```

2. Before you can publish the package, you need to add the PyPI repository to Poetry if you haven't done so already. You can do this with the following command:

```bash
poetry config repositories.pypi https://upload.pypi.org/legacy/
```

4. Now you can publish your package to PyPI. You'll be prompted to enter your PyPI username and password:

```bash
poetry publish -r pypi
```

## Checklist

- [x] I have added dependencies to the `pyproject.toml` file
- [x] I have updated the `poetry.lock` file
- [x] I have updated the `requirements.txt` file (till 2024-06-01)
- [ ] New and existing integration notebooks pass locally with my changes
- [x] Pre-Commit checks are passing (locally and remotely)
- [x] Title of your PR / MR corresponds to the required format
- [x] I have added Google Colab support for the newly added notebooks.
- [ ] I have added a new notebook that tests the new feature
- [x] I ran `ruff check . --fix-only` to appease the lint gods
- [x] Commit message follows required format {label}(dspy): {message}
